borg/repository.py

```diff
@@ -207,14 +207,18 @@ def open_index(self, transaction_id):
 
     def prepare_txn(self, transaction_id, do_cleanup=True):
         self._active_txn = True
-        try:
-            self.lock.upgrade()
-        except (LockError, LockErrorT):
-            # if upgrading the lock to exclusive fails, we do not have an
-            # active transaction. this is important for "serve" mode, where
-            # the repository instance lives on - even if exceptions happened.
-            self._active_txn = False
-            raise
+        if not self.lock.got_exclusive_lock():
+            # this is either an old client talking to a new server (expecting lock upgrade) or
```

Comment:
If we want this to be enforcing that upgrades are only used by old code and by index replay we could do a
    if self.exclusive is not None:
        assert("like in your old version")

and change
* the default in remote to be None.
* in __enter__ use True if self.exclusive else False
* change the replay code to disable this check by setting self.exclusive to None